introspection-engine/connectors/sql-introspection-connector/tests/test_harness/introspection_database.rs

```diff
@@ -0,0 +1,220 @@
+    where
+        F: FnOnce(&mut dyn Queryable) -> T,
+    {
+        match self.conn {
+            PostgresConnection::Single(ref mutex) => f(mutex.lock().unwrap().deref_mut()),
+            PostgresConnection::Pooled(ref pool) => {
+                let mut conn = pool.get().unwrap();
+                f(conn.deref_mut())
+            }
+        }
+    }
+}
+
+impl IntrospectionDatabase for PostgreSql {
+    fn execute(&self, _: &str, q: Query) -> prisma_query::Result<Option<Id>> {
+        self.with_connection(|conn| conn.execute(q))
+    }
+
+    fn query(&self, _: &str, q: Query) -> prisma_query::Result<ResultSet> {
+        self.with_connection(|conn| conn.query(q))
+    }
+
+    fn query_raw(&self, _: &str, sql: &str, params: &[ParameterizedValue]) -> prisma_query::Result<ResultSet> {
+        self.with_connection(|conn| conn.query_raw(sql, params))
+    }
+
+    fn execute_raw(&self, _: &str, sql: &str, params: &[ParameterizedValue]) -> prisma_query::Result<u64> {
+        self.with_connection(|conn| conn.execute_raw(sql, params))
+    }
+}
+
+enum MysqlConnection {
+    Pooled(MysqlPool),
+    Single(Mutex<connector::Mysql>),
+}
+
+pub struct Mysql {
+    conn: MysqlConnection,
+}
+
+impl Mysql {
+    pub fn new(params: MysqlParams, pooled: bool) -> prisma_query::Result<Self> {
+        let conn = if pooled {
+            let manager = PrismaConnectionManager::mysql(params.config);
+
+            let pool = r2d2::Pool::builder()
+                .connection_timeout(Duration::from_millis(1500))
+                .max_size(2)
+                .test_on_check_out(false)
+                .build(manager)?;
+
+            MysqlConnection::Pooled(pool)
+        } else {
+            let conn = connector::Mysql::from_params(params)?;
+            MysqlConnection::Single(Mutex::new(conn))
+        };
+
+        Ok(Self { conn })
+    }
+
+    fn with_connection<F, T>(&self, f: F) -> T
+    where
+        F: FnOnce(&mut dyn Queryable) -> T,
+    {
+        match self.conn {
+            MysqlConnection::Single(ref mutex) => f(mutex.lock().unwrap().deref_mut()),
+            MysqlConnection::Pooled(ref pool) => {
+                let mut conn = pool.get().unwrap();
+                f(conn.deref_mut())
+            }
+        }
+    }
+}
+
+impl IntrospectionDatabase for Mysql {
+    fn execute(&self, _: &str, q: Query) -> prisma_query::Result<Option<Id>> {
+        self.with_connection(|conn| conn.execute(q))
+    }
+
+    fn query(&self, _: &str, q: Query) -> prisma_query::Result<ResultSet> {
+        self.with_connection(|conn| conn.query(q))
+    }
+
+    fn query_raw(&self, _: &str, sql: &str, params: &[ParameterizedValue]) -> prisma_query::Result<ResultSet> {
+        self.with_connection(|conn| conn.query_raw(sql, params))
+    }
+
+    fn execute_raw(&self, _: &str, sql: &str, params: &[ParameterizedValue]) -> prisma_query::Result<u64> {
+        self.with_connection(|conn| conn.execute_raw(sql, params))
+    }
+}
```

Comment:
So this is the code that is duplicated from the migrations engine. I think we could take the opportunity of the switch to the async prisma-query to make a proper crate for that code, since it will have to be mostly rewritten anyway.